addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/HeaderBasedSessionManagementMethodType.java

```diff
@@ -0,0 +1,454 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.db.DatabaseException;
+import org.parosproxy.paros.db.RecordContext;
+import org.parosproxy.paros.extension.ExtensionHook;
+import org.parosproxy.paros.model.Session;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.view.View;
+import org.zaproxy.zap.extension.api.ApiDynamicActionImplementor;
+import org.zaproxy.zap.extension.api.ApiException;
+import org.zaproxy.zap.extension.api.ApiResponse;
+import org.zaproxy.zap.extension.api.ApiResponseElement;
+import org.zaproxy.zap.extension.script.ScriptVars;
+import org.zaproxy.zap.extension.sessions.ExtensionSessionManagement;
+import org.zaproxy.zap.extension.sessions.SessionManagementAPI;
+import org.zaproxy.zap.model.Context;
+import org.zaproxy.zap.session.AbstractSessionManagementMethodOptionsPanel;
+import org.zaproxy.zap.session.SessionManagementMethod;
+import org.zaproxy.zap.session.SessionManagementMethodType;
+import org.zaproxy.zap.session.WebSession;
+import org.zaproxy.zap.utils.ApiUtils;
+import org.zaproxy.zap.utils.EncodingUtils;
+import org.zaproxy.zap.utils.Pair;
+import org.zaproxy.zap.view.LayoutHelper;
+
+/**
+ * The type corresponding to a {@link org.zaproxy.zap.session.SessionManagementMethod} for web
+ * applications that use Header based Authentication.
+ */
+public class HeaderBasedSessionManagementMethodType extends SessionManagementMethodType {
+
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_HEADER =
+            ExtensionSessionManagement.CONTEXT_CONFIG_SESSION + ".headers";
+
+    private static final int METHOD_IDENTIFIER = 3;
+
+    private static final String API_METHOD_NAME = "headerBasedSessionManagement";
+
+    private static final Logger LOGGER =
+            LogManager.getLogger(HeaderBasedSessionManagementMethodType.class);
+
+    public static class HeaderBasedSessionManagementMethod implements SessionManagementMethod {
+
+        private List<Pair<String, String>> headerConfigs = new ArrayList<>();
+        private static Map<String, String> envVars = System.getenv();
+
+        @Override
+        public boolean isConfigured() {
+            return true;
+        }
+
+        @Override
+        public SessionManagementMethodType getType() {
+            return new HeaderBasedSessionManagementMethodType();
+        }
+
+        protected static String replaceTokens(String text, Map<String, String> tokens) {
+            // Based on https://stackoverflow.com/a/959776/1509834
+            Pattern pattern = Pattern.compile("\\{%(.+?)\\%}");
+            Matcher matcher = pattern.matcher(text);
+            StringBuilder builder = new StringBuilder();
+            int i = 0;
+            while (matcher.find()) {
+                String replacement = tokens.get(matcher.group(1));
+                builder.append(text.substring(i, matcher.start()));
```

Comment:
We should use the `Matcher` methods to append, the Java version we target supports `StringBuilder`.

Reply:
Oh nice - I didnt know about these 😁